src/prebundle-modules-plugin.ts

```diff
@@ -0,0 +1,180 @@
+import esbuildCommonjsPlugin from '@chialab/esbuild-plugin-commonjs';
+import { stringifyProcessEnvs } from '@storybook/core-common';
+import type { Plugin } from '@web/dev-server-core';
+import { build } from 'esbuild';
+import { join } from 'path';
+import { getNodeModuleDir } from './get-node-module-dir';
+
+export const PREBUNDLED_MODULES_DIR = 'node_modules/.prebundled_modules';
+
+export function prebundleModulesPlugin(env: Record<string, string>): Plugin {
+  const modulePaths: Record<string, string> = {};
+
+  return {
+    name: 'prebundle-modules',
+
+    async serverStart() {
+      const modules = getModules();
+
+      for (const module of modules) {
+        modulePaths[module] =
+          '/' + join(PREBUNDLED_MODULES_DIR, module.endsWith('.js') ? module : `${module}.js`);
+      }
+
+      await build({
+        entryPoints: modules,
+        outdir: PREBUNDLED_MODULES_DIR,
+        bundle: true,
+        format: 'esm',
+        splitting: true,
+        sourcemap: true,
+        alias: {
+          assert: require.resolve('browser-assert'),
+          lodash: getNodeModuleDir('lodash-es'), // more optimal, but also solves esbuild incorrectly compiling lodash/_nodeUtil
+          path: require.resolve('path-browserify'),
+        },
+        define: {
+          ...stringifyProcessEnvs(env),
+        },
+        plugins: [esbuildCommonjsPlugin()],
+      });
+    },
+
+    async resolveImport({ source }) {
+      return modulePaths[source];
+    },
+  };
+}
+
+function getModules() {
+  const include = CANDIDATES.filter(id => {
+    try {
+      require.resolve(id, { paths: [process.cwd()] });
+      return true;
+    } catch (e) {
+      return false;
+    }
+  });
+  return include;
+}
+
+// taken from https://github.com/storybookjs/storybook/blob/v7.0.9/code/lib/builder-vite/src/optimizeDeps.ts#L7
+export const CANDIDATES = [
+  '@base2/pretty-print-object',
+  '@emotion/core',
+  '@emotion/is-prop-valid',
+  '@emotion/styled',
+  '@mdx-js/react',
+  '@storybook/addon-docs > acorn-jsx',
```

Comment:
this is still TODO, but I wanna check first with builder-vite maintainers the use-case here